### DISCUSSION TITLE
Can consumers of this library have access to testing utilities?

### DISCUSSION THREAD

**Discussion by stephenbelyea** [react-querybuilder/react-querybuilder#539] [Q&A] [answered]
Hello! 

Would it be possible for developers who are making use of React Query Builder could have exported access to some of your team's internal testing utilities, like `ruleTestUtils`:

https://github.com/react-querybuilder/react-querybuilder/blob/main/packages/react-querybuilder/genericTests/ruleTestUtils.tsx

When customizing the components used in our own `QueryBuilder` implementation, it would be a huge help to have an easy one-liner to generate all the default and required props, rather than manually building up a mock.

For context, my specific use case involves wrapping the `Rule` component with a container which offers an `id` and `tabIndex` value, allowing us to set the user's focus to the component when a new rule is created. This is a key accessibility enhancement that is currently missing from the library!

Thanks!

**Comment by jakeboone02** [comment, marked as answer]
Hmmm...while I do understand the use case, I'm reluctant to expose those utilities and fixtures because they're subject to change at the whim of a sole developer. I wouldn't really want to be forced (by semver) to release a higher-level version (minor instead of patch, or major instead of minor) just because some of the test fixtures changed.

I'll consider it, or maybe an official test-utils package or something. In the meantime I would suggest copy-pasting everything you need from the [genericTests folder](https://github.com/react-querybuilder/react-querybuilder/tree/main/packages/react-querybuilder/genericTests). I have no plans to change anything there in the foreseeable future, so you should be safe for a good while.

Out of curiosity, how are you generating the `id` and `tabIndex` for your `Rule` wrapper? If possible, is it something you'd like to see baked into the library?

**Reply by stephenbelyea** [reply]
> because they're subject to change at the whim of a sole developer

That is completely fair and I definitely understand not wanting to lock down any personal-ish dev tools within a codebase! A separate test-util package would be pretty huge and I imagine could be helpful for lots of folks!

Right now, I'm using the `id` passed with each `RuleProps` (and for groups as well), wrapping and re-exporting the `Rule` component with a `div`, and adding the `id` and `tabIndex` to that container. Since we don't want the container to receive focus when the user is Tabbing through the interface, `tabIndex` only ever needs to be `-1` (allows it to receive focus programmatically only). From there, I'm setting a global state variable to the new rule's `id` within the `onAddRule` function and then manually setting the focus to the new rule's container when `useEffect` picks up the state change for the ID value. This means that the user's next Tab press lands them directly in the new rule's first input, rather than still being left at the top of the group or rule stack.

And yes - any accessibility enhancement would be fantastic within the library! I'm currently working on a few for our usage of React Query Builder; mostly to do with managing the user's keyboard focus as they create, remove, or re-order rules and groups.

**Reply by jakeboone02** [reply]
I don't know if I ever mentioned that I had an issue filed for this: #552
